_posts/2019-09-19-sorting-a-table-with-live-view-live-links.md

```diff
@@ -0,0 +1,197 @@
+<table>
+  <th><%= live_link "Name", to: Routes.live_path(@socket, CourseConductorWeb.CohortsLive, %{sort_by: "name"}) %></th>
+  ...
+</table>
+```
+
+The `live_link/2` function generates a live link for HTML5 pushState based navigation *without* page reloads.
+
+With the help of the the `Routes.live_path` helper, we're generating the following live link: `"/cohorts?sort_by=name"`. Since this route belongs to the `CohortsLive` live view that we've already mounted, _and_ since that live view is defined in our router (as opposed to rendered from a controller action), this means we will invoke our existing live view's `handle_params/3` function _without mounting a new LiveView_. Pretty cool!
+
+Let's take a look at how we can implement a `handle_params/3` function now.
+
+## Implementing `handle_params/3`
+
+The `handle_params/3` callback is invoked under two circumstances.
+* After `mount/2` is called (i.e. when the live view first renders)
+* When a live navigation event, like a live link click, occurs. This second circumstance only triggers this callback when, as described above, the live view we are linking to is the same live view we are currently on _and_ the LiveView is defined in the router.  
+
+`handle_params/3` receives three arguments:
+* The query parameters
+* The requested url
+* The socket
+
+We can use `handle_params/3` to update socket state and therefore trigger a server re-render of the template.
+
+Given that `handle_params/3` will be invoked by our live view whenever our `"Name"` live link is clicked, we need to implement this function in our live view to match and act on the `sort_by` params our live link will send.
+
+Assuming we have the following live view that mounts and renders a list of cohorts:
+
+```elixir
+# lib/course_conductor_web/live/cohorts_live.ex
+
+defmodule CourseConductorWeb.CohortsLive do
+  use Phoenix.LiveView
+
+  def render(assigns) do
+    Phoenix.View.render(CourseConductorWeb.CohortView, "index.html", assigns)
+  end
+
+  def mount(_, socket) do
+    cohorts = Cohort.all_cohorts()
+    {:ok, assign(socket, cohorts: cohorts)}
+  end
+end
+```
+
+We'll implement our `handle_params/3` function like this:
+
+```elixir
+#  lib/course_conductor_web/live/cohorts_live.ex
+
+def handle_params(%{"sort_by" => sort_by}, _uri, socket) do
+  case sort_by do
+    sort_by
+    when sort_by in ~w(name) ->
+      {:noreply, assign(socket, cohorts: sort_cohorts(socket.assigns.cohorts, sort_by))}
+
+    _ ->
+      {:noreply, socket}
+  end
+end
+
+
+def handle_params(_params, _uri, socket) do
+  {:noreply, socket}
+end
+
+def sort_cohorts(cohort, "name") do
+  Enum.sort_by(cohorts, fn cohort -> cohort.name end)
+end
+```
+
+Note that we've included a "catch-all" version of the `handle_params/3` function that will be invoked if someone navigates to `/cohorts` and includes query params that do not match the `"sort_by"` param that we care about. If our live view recieves such a request, it will not update state.
```

Comment:
```suggestion
Note that we've included a "catch-all" version of the `handle_params/3` function that will be invoked if someone navigates to `/cohorts` and includes query params that do not match the `"sort_by"` param that we care about. If our live view receives such a request, it will not update state.
```